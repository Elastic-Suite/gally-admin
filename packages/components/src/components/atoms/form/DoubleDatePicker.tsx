import React, { ReactNode } from 'react'
import { Box, FormHelperText, Grid, InputLabel } from '@mui/material'
import { DateValidationError } from '@mui/x-date-pickers/internals/hooks/validation/useDateValidation'
import { useTranslation } from 'next-i18next'
import { styled } from '@mui/system'
import dayjs, { Dayjs } from 'dayjs'

import IonIcon from '../IonIcon/IonIcon'

import DatePicker, { IDatePickerProps } from './DatePicker'
import FormControl from './FormControl'
import InfoTooltip from './InfoTooltip'
import { LocalizationProvider } from '@mui/x-date-pickers'
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs'

const CustomBox = styled(Box)(() => ({
  fontWeight: 400,
  fontFamily: 'var(--gally-font)',
}))

export interface IDoubleDatePickerValues {
  from: Dayjs | string | null
  to: Dayjs | string | null
}
export interface IDoubleDatePickerErrors {
  from: DateValidationError
  to: DateValidationError
}

export interface IDoubleDatePickerProps
  extends Omit<IDatePickerProps, 'value' | 'onChange' | 'onError'> {
  value?: IDoubleDatePickerValues
  onChange?: (values: IDoubleDatePickerValues) => void
  onError?: (reasons: IDoubleDatePickerErrors) => void
  error?: boolean
  errors?: IDoubleDatePickerErrors
  fullWidth?: boolean
  infoTooltip?: string
  label?: ReactNode
  margin?: 'none' | 'dense' | 'normal'
  helperText?: ReactNode
  helperIcon?: string
  stringFormat?: boolean
}

function DoubleDatePicker(props: IDoubleDatePickerProps): JSX.Element {
  const {
    value,
    error,
    errors,
    fullWidth,
    helperText,
    helperIcon,
    id,
    infoTooltip,
    inputProps,
    label,
    margin,
    onChange,
    onError,
    required,
    stringFormat = true,
    ...args
  } = props
  const { t } = useTranslation('common')

  function stringToDayJs(date: string): Dayjs | null {
    if (!date) {
      return null
    }
    return dayjs(date)
  }

  function onChangeFrom(date: Dayjs | Record<string, string>): void {
    const newFormatDate = stringFormat
      ? new Date((date as Record<string, string>).$d).toLocaleDateString(
          'en-EN'
        ) // TODO change en-EN with global variable of I18N
      : (date as Dayjs)
    onChange({ ...value, from: newFormatDate })
  }

  function onChangeTo(date: Dayjs | null): void {
    onChange({ ...value, to: date })
  }

  function onErrorFrom(reason: DateValidationError): void {
    onError({ ...errors, from: reason })
  }

  function onErrorTo(reason: DateValidationError): void {
    onError({ ...errors, to: reason })
  }

  return (
    <FormControl error={error} fullWidth={fullWidth} margin={margin}>
      {Boolean(label || infoTooltip) && (
        <InputLabel shrink htmlFor={id} required={required}>
          {label}
          {infoTooltip ? <InfoTooltip title={infoTooltip} /> : null}
        </InputLabel>
      )}
      <Grid
        justifyContent="flex-start"
        alignItems="center"
        container
        sx={{ marginTop: label ? '24px' : '0px' }}
      >
        <CustomBox sx={{ paddingRight: '20px' }}> {t('form.from')} </CustomBox>
        <LocalizationProvider dateAdapter={AdapterDayjs}>
          <Grid item xs>
            <DatePicker
              {...args}
              error={error}
              fullWidth={fullWidth}
              inputProps={inputProps}
              value={
                stringFormat
                  ? stringToDayJs(value?.from as string)
                  : (value?.from as Dayjs)
              }
              onChange={onChangeFrom}
              onError={onErrorFrom}
            />
          </Grid>
          <CustomBox sx={{ paddingRight: '20px', paddingLeft: '20px' }}>
            {t('form.to')}
          </CustomBox>
          <Grid item xs>
            <DatePicker
              {...args}
              error={error}
              fullWidth={fullWidth}
              inputProps={inputProps}
              value={
                stringFormat
                  ? stringToDayJs(value?.to as string)
                  : (value?.to as Dayjs)
              }
              onChange={onChangeTo}
              onError={onErrorTo}
            />
          </Grid>
        </LocalizationProvider>
      </Grid>
      {Boolean(helperText) && (
        <FormHelperText>
          {Boolean(helperIcon) && (
            <IonIcon
              name={helperIcon}
              style={{ fontSize: 18, marginRight: 2 }}
            />
          )}
          {helperText}
        </FormHelperText>
      )}
    </FormControl>
  )
}

export default DoubleDatePicker
