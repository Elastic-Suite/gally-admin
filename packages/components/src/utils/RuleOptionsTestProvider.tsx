import React, { ReactNode } from 'react'
import {
  IRuleEngineOperators,
  RuleAttributeType,
  getIri,
} from '@elastic-suite/gally-admin-shared'

import ruleEngineOperator from '../../public/mocks/rule_engine_operators.json'
import sourceFields from '../../public/mocks/source_fields.json'
import sourceFieldLabels from '../../public/mocks/source_field_labels.json'

import RuleOptionsProvider from '../components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(getIri('source_fields', field.id))
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator as IRuleEngineOperators}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
