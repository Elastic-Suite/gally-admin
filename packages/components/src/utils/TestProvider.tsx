import React, { ReactNode } from 'react'

import catalogs from '../../public/mocks/catalog.json'
import metadata from '../../public/mocks/metadata.json'

import { catalogContext, metadataContext } from '../contexts'
import { useCatalogs } from '../hooks'

import OptionsProvider from '../components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { children } = props
  const contextValueCatalog = useCatalogs(catalogs['hydra:member'])
  const contextValueMetadata = metadata['hydra:member']

  return (
    <OptionsProvider>
      <catalogContext.Provider value={contextValueCatalog}>
        <metadataContext.Provider value={contextValueMetadata}>
          {children}
        </metadataContext.Provider>
      </catalogContext.Provider>
    </OptionsProvider>
  )
}

export default TestProvider
