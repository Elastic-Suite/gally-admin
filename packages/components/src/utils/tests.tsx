import React, { ReactNode } from 'react'
import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { Bundle, api } from '@elastic-suite/gally-admin-shared'

import { AppStore, RootState, setupStore } from '../store'

import AppProvider from '../components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: {
      'base_url/media': 'https://localhost/media/catalog/product/',
    },
    metadata: [
      {
        '@id': '/metadata/3',
        '@type': 'Metadata',
        id: 3,
        entity: 'product',
        sourceFields: ['/source_fields/217', '/source_fields/253'],
      },
      {
        '@id': '/metadata/4',
        '@type': 'Metadata',
        id: 4,
        entity: 'category',
        sourceFields: ['/source_fields/254', '/source_fields/286'],
      },
    ],
  },
  user: {
    requestedPath: '',
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
