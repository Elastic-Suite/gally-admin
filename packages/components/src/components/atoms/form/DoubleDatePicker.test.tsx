import React from 'react'

import { renderWithProviders } from '../../../utils/tests'

import DoubleDatePicker from './DoubleDatePicker'
import { screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

describe('DoubleDatePicker', () => {
  it('sould match snapshot', () => {
    const { container } = renderWithProviders(
      <DoubleDatePicker
        label="Label"
        value={{
          toDate: new Date('Fri, 12 Apr 2024 12:40:52 GMT'),
          fromDate: new Date('Fri, 12 Apr 2024 12:40:52 GMT'),
        }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should display error with showError prop when the field is required with missing value', () => {
    renderWithProviders(
      <DoubleDatePicker
        label="Label"
        required
        showError
        value={{ toDate: null, fromDate: null }}
      />
    )
    expect(screen.getByText('formError.valueMissing')).toBeInTheDocument()
  })

  it('should display error with showError prop when the field have an invalid value', () => {
    renderWithProviders(
      <DoubleDatePicker
        label="Label"
        required
        showError
        value={{ toDate: 'invalid value', fromDate: new Date() }}
      />
    )
    expect(screen.getByText('formError.invalidDate')).toBeInTheDocument()
  })

  it('should display error with showError prop when the first date is superior to the second date', () => {
    renderWithProviders(
      <DoubleDatePicker
        label="Label"
        showError
        value={{ toDate: new Date(1680605116), fromDate: new Date() }}
      />
    )
    expect(
      screen.getByText('formError.doubleDatePickerRange')
    ).toBeInTheDocument()
  })

  it('should not display error without showError prop when the field is required with missing value', () => {
    renderWithProviders(
      <DoubleDatePicker
        label="Label"
        required
        value={{ toDate: null, fromDate: null }}
      />
    )
    expect(screen.queryByText('formError.valueMissing')).not.toBeInTheDocument()
  })

  it('should not display error with showError prop when the field is required and disabled with missing value', () => {
    renderWithProviders(
      <DoubleDatePicker
        label="Label"
        required
        showError
        disabled
        value={{ toDate: null, fromDate: null }}
      />
    )
    expect(screen.queryByText('formError.valueMissing')).not.toBeInTheDocument()
  })

  it('shoud display error with showError prop when the field has additional validation rules', () => {
    renderWithProviders(
      <DoubleDatePicker
        label="Label"
        showError
        required
        value={{ toDate: null, fromDate: null }}
        additionalValidator={(): string => {
          return 'erreur'
        }}
      />
    )
    expect(screen.getByText('formError.erreur')).toBeInTheDocument()
  })

  it('should display a replacement error message with replacementErrorsMessages prop when the field has an error', () => {
    renderWithProviders(
      <DoubleDatePicker
        label="Label"
        required
        showError
        value={{ toDate: null, fromDate: null }}
        replacementErrorsMessages={{
          valueMissing: 'customError',
        }}
      />
    )
    expect(screen.getByText('formError.customError')).toBeInTheDocument()
  })
})
