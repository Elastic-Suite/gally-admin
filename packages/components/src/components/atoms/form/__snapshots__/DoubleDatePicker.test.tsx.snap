// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoubleDatePicker sould match snapshot 1`] = `
<div>
  <div
    class="css-vaijrj"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-colorPrimary css-1j7os2p-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
    >
      Label
    </label>
    <div
      class="MuiGrid-root MuiGrid-container css-1v2wnak-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-1i75mai"
      >
         
        form.from
         
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
        >
          <div
            class="InputText__Wrapper css-17fp4jk"
          >
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-1jf161a-MuiInputBase-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInputBase-inputAdornedEnd css-ngazp5-MuiInputBase-input"
                data-testid="undefinedStart"
                placeholder="date.placeholder"
                type="tel"
                value="04/12/2024"
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <button
                  aria-label="Choose date, selected date is Apr 12, 2024"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <ion-icon
                    name="calendar-outline"
                    style="font-size: 18px; padding: 0px;"
                  />
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-ow4s8t"
      >
        form.to
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
        >
          <div
            class="InputText__Wrapper css-17fp4jk"
          >
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-1jf161a-MuiInputBase-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInputBase-inputAdornedEnd css-ngazp5-MuiInputBase-input"
                data-testid="undefinedEnd"
                placeholder="date.placeholder"
                type="tel"
                value="04/12/2024"
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <button
                  aria-label="Choose date, selected date is Apr 12, 2024"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <ion-icon
                    name="calendar-outline"
                    style="font-size: 18px; padding: 0px;"
                  />
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
